File: typescript/packages/core/src/core/langfuse/read.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { LangfuseAccessor } from '../../accessor/langfuse.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import type { PathSpec } from '../../types.ts'
import { fetchDatasetItems, fetchDatasetRuns, fetchPrompt, fetchTrace } from './_client.ts'
import { stripSlash } from '../../util/slash.ts'

const ENC = new TextEncoder()

function enoent(path: string): Error {
  const err = new Error(`ENOENT: ${path}`) as Error & { code: string }
  err.code = 'ENOENT'
  return err
}

function toJsonBytes(data: unknown): Uint8Array {
  return ENC.encode(JSON.stringify(data, null, 2))
}

function toJsonlBytes(items: readonly Record<string, unknown>[]): Uint8Array {
  if (items.length === 0) return new Uint8Array()
  const text = items.map((item) => JSON.stringify(item)).join('\n') + '\n'
  return ENC.encode(text)
}

export async function read(
  accessor: LangfuseAccessor,
  path: PathSpec,
  _index?: IndexCacheStore,
): Promise<Uint8Array> {
  const prefix = path.prefix
  let p = path.original
  if (prefix !== '' && p.startsWith(prefix)) {
    p = p.slice(prefix.length) || '/'
  }
  const key = stripSlash(p)
  if (key === '') throw enoent(p)
  const parts = key.split('/')
  for (const part of parts) {
    if (part.startsWith('.')) throw enoent(p)
  }

  if (parts[0] === 'traces' && parts.length === 2 && (parts[1] ?? '').endsWith('.json')) {
    const traceId = (parts[1] ?? '').slice(0, -'.json'.length)
    const data = await fetchTrace(accessor.transport, traceId)
    return toJsonBytes(data)
  }

  if (parts[0] === 'sessions' && parts.length === 3 && (parts[2] ?? '').endsWith('.json')) {
    const traceId = (parts[2] ?? '').slice(0, -'.json'.length)
    const data = await fetchTrace(accessor.transport, traceId)
    return toJsonBytes(data)
  }

  if (parts[0] === 'prompts' && parts.length === 3 && (parts[2] ?? '').endsWith('.json')) {
    const promptName = parts[1] ?? ''
    const versionStr = (parts[2] ?? '').slice(0, -'.json'.length)
    const version = Number.parseInt(versionStr, 10)
    if (Number.isNaN(version)) throw enoent(p)
    const data = await fetchPrompt(accessor.transport, promptName, version)
    return toJsonBytes(data)
  }

  if (parts[0] === 'datasets' && parts.length === 3 && parts[2] === 'items.jsonl') {
    const datasetName = parts[1] ?? ''
    const items = await fetchDatasetItems(accessor.transport, datasetName)
    return toJsonlBytes(items)
  }

  if (
    parts[0] === 'datasets' &&
    parts.length === 4 &&
    parts[2] === 'runs' &&
    (parts[3] ?? '').endsWith('.jsonl')
  ) {
    const datasetName = parts[1] ?? ''
    const runName = (parts[3] ?? '').slice(0, -'.jsonl'.length)
    const runs = await fetchDatasetRuns(accessor.transport, datasetName)
    const matched = runs.filter((r) => r.name === runName)
    const first = matched[0]
    if (first === undefined) throw enoent(p)
    return toJsonBytes(first)
  }

  throw enoent(p)
}

File: python/mirage/core/langfuse/read.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json

from mirage.accessor.langfuse import LangfuseAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.langfuse._client import (fetch_dataset_items,
                                          fetch_dataset_runs, fetch_prompt,
                                          fetch_trace)
from mirage.types import PathSpec
from mirage.utils.errors import enoent


def _json_bytes(data: dict) -> bytes:
    return json.dumps(data, ensure_ascii=False, indent=2).encode()


def _jsonl_bytes(items: list[dict]) -> bytes:
    if not items:
        return b""
    lines = [json.dumps(item, ensure_ascii=False) for item in items]
    return ("\n".join(lines) + "\n").encode()


async def read(
    accessor: LangfuseAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> bytes:
    """Read a file as bytes.

    Args:
        accessor (LangfuseAccessor): langfuse accessor.
        path (str): resource-relative path.
        index (IndexCacheStore | None): index cache.
        prefix (str): mount prefix for virtual index keys.
    """
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    virtual = path.original
    if isinstance(path, PathSpec):
        prefix = path.prefix
        path = path.original

    if prefix and path.startswith(prefix):
        rest = path[len(prefix):]
        if prefix.endswith("/") or rest == "" or rest.startswith("/"):
            path = rest or "/"
    key = path.strip("/")

    if any(p.startswith(".") for p in key.split("/")):
        raise enoent(virtual)

    parts = key.split("/")

    if parts[0] == "traces" and len(parts) == 2 and parts[1].endswith(".json"):
        trace_id = parts[1].removesuffix(".json")
        data = await fetch_trace(accessor.api, trace_id)
        return _json_bytes(data)

    if (parts[0] == "sessions" and len(parts) == 3
            and parts[2].endswith(".json")):
        trace_id = parts[2].removesuffix(".json")
        data = await fetch_trace(accessor.api, trace_id)
        return _json_bytes(data)

    if (parts[0] == "prompts" and len(parts) == 3
            and parts[2].endswith(".json")):
        prompt_name = parts[1]
        version = int(parts[2].removesuffix(".json"))
        data = await fetch_prompt(accessor.api, prompt_name, version)
        return _json_bytes(data)

    if (parts[0] == "datasets" and len(parts) == 3
            and parts[2] == "items.jsonl"):
        dataset_name = parts[1]
        items = await fetch_dataset_items(accessor.api, dataset_name)
        return _jsonl_bytes(items)

    if (parts[0] == "datasets" and len(parts) == 4 and parts[2] == "runs"
            and parts[3].endswith(".jsonl")):
        dataset_name = parts[1]
        run_name = parts[3].removesuffix(".jsonl")
        runs = await fetch_dataset_runs(accessor.api, dataset_name)
        matched = [r for r in runs if r.get("name") == run_name]
        if not matched:
            raise enoent(virtual)
        return _json_bytes(matched[0])

    raise enoent(virtual)
